libfwupd/fwupd-bios-attr.c

```diff
@@ -0,0 +1,878 @@
+	if (fwupd_bios_attr_has_possible_value(self, possible_value))
+		return;
+	g_ptr_array_add(priv->possible_values, g_strdup(possible_value));
+}
+
+/**
+ * fwupd_bios_attr_get_name:
+ * @self: a #FwupdBiosAttr
+ *
+ * Gets the attribute name.
+ *
+ * Returns: the attribute name, or %NULL if unset
+ *
+ * Since: 1.8.4
+ **/
+const gchar *
+fwupd_bios_attr_get_name(FwupdBiosAttr *self)
+{
+	FwupdBiosAttrPrivate *priv = GET_PRIVATE(self);
+	g_return_val_if_fail(FWUPD_IS_BIOS_ATTR(self), NULL);
+	return priv->name;
+}
+
+/**
+ * fwupd_bios_attr_get_path:
+ * @self: a #FwupdBiosAttr
+ *
+ * Gets the path for the driver providing the attribute.
+ *
+ * Returns: (nullable): the driver, or %NULL if unfound
+ *
+ * Since: 1.8.4
+ **/
+const gchar *
+fwupd_bios_attr_get_path(FwupdBiosAttr *self)
+{
+	FwupdBiosAttrPrivate *priv = GET_PRIVATE(self);
+	g_return_val_if_fail(FWUPD_IS_BIOS_ATTR(self), NULL);
+	return priv->path;
+}
+
+/**
+ * fwupd_bios_attr_get_description:
+ * @self: a #FwupdBiosAttr
+ *
+ * Gets the attribute description which is provided by some drivers to explain
+ * what they change.
+ *
+ * Returns: the attribute description, or %NULL if unset
+ *
+ * Since: 1.8.4
+ **/
+const gchar *
+fwupd_bios_attr_get_description(FwupdBiosAttr *self)
+{
+	FwupdBiosAttrPrivate *priv = GET_PRIVATE(self);
+	g_return_val_if_fail(FWUPD_IS_BIOS_ATTR(self), NULL);
+	return priv->description;
+}
+
+static gchar *
+fwupd_bios_attr_build(FwupdBiosAttr *self, GError **error)
+{
+	FwupdBiosAttrPrivate *priv = GET_PRIVATE(self);
+	g_autofree gchar *tmp = NULL;
+	const gchar *current_value;
+
+	g_return_val_if_fail(FWUPD_IS_BIOS_ATTR(self), NULL);
+
+	current_value = g_strcmp0(priv->name, FWUPD_BIOS_ATTR_PENDING_REBOOT) != 0 ? "current_value"
+										   : NULL;
+	tmp = g_build_filename(priv->path, priv->name, current_value, NULL);
+	if (!g_file_test(tmp, G_FILE_TEST_EXISTS)) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_NOT_FOUND,
+			    "failed to find attribute %s",
+			    tmp);
+		return NULL;
+	}
+
+	return g_steal_pointer(&tmp);
+}
+
+/**
+ * fwupd_bios_attr_get_current_value:
+ * @self: a #FwupdBiosAttr
+ * @error: (nullable): optional return location for an error
+ *
+ * Gets the string representation of the current_value stored in an attribute
+ * from the kernel.
+ * This value is not cached and will be fetched from the kernel on demand.
+ *
+ * Since: 1.8.4
+ **/
+gchar *
+fwupd_bios_attr_get_current_value(FwupdBiosAttr *self, GError **error)
+{
+	g_autofree gchar *path = NULL;
+	g_autofree gchar *data = NULL;
+	gsize len;
+
+	path = fwupd_bios_attr_build(self, error);
+	if (path == NULL)
+		return NULL;
+
+	g_return_val_if_fail(FWUPD_IS_BIOS_ATTR(self), NULL);
+
+	if (!g_file_get_contents(path, &data, &len, error))
+		return NULL;
+
+	g_strchomp(data);
+	return g_steal_pointer(&data);
+}
+
+/**
+ * fwupd_bios_attr_set_current_value:
+ * @self: a #FwupdBiosAttr
+ * @value: The string to set an attribute to
```

Comment:
maybe add `(element-type utf8)` to these?

Reply:
fixed